freedomofpress/securedrop
# Creates Xenial staging scenario

## Status

Ready for review.

## Description of Changes

Closes #3206. 

Changes proposed in this pull request:

* Creates new Xenial-based staging scenarios for libvirt and virtualbox (not implemented for Qubes yet).
* Tweaks Ansible logic to support overrides of apt repo channels, so we can install "trusty" packages on "xenial" hosts for debugging
* Slightly modifies dependencies for `securedrop-app-code` _when Xenial support is enabled_. 
* Created new Makefile targets `staging-xenial` and `build-debs-xenial`. 
* Wrote docs to explain the above

The install flow is not perfect here—errors _will_ be encountered periodically. Still, the goal here is to provide a foothold for developers to start fixing these errors in a way that preserves Trusty compatibility. As such, 

## Testing
1. Confirm `make build-debs` has all tests passing (packaging behavior should not change).
2. Confirm `make build-debs-xenial` emits a `securedrop-app-code` package that has modified dependencies; specifically, `apache2` is now explicitly included, and `apache2-mpm-worker` is omitted.
3. Run `make staging-xenial` and report how far the install gets. It won't get all the way to the end; at the very least, the `apache2-mpm-worker` is still re-included in the Ansible logic (which can probably be removed, given the deb package dependencies). 
4. Re-run `make staging-xenial` and see if the install gets further than last time.

Review the docs presented and comment if any additional changes should be made. Tried to cross-link with prior research efforts to begin centralizing the upcoming fix-it tasks as we push toward Xenial support.

## Deployment

Shouldn't be, but requires testing. Some tweaks (off-by-default) were made to packaging logic, and minor adjustments to the Ansible logic. All were intended to preserve the existing default behavior when run against Trusty.

## Checklist

### If you made changes to the server application code:


### If you made changes to `securedrop-admin`:

- N/A

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass (Yes, changes were made here, but the config tests don't support Xenial yet; I suggest we defer to a future ticket)

### If you made non-trivial code changes:


### If you made changes to documentation:
